File: front-end/src/services/api/reports.ts
/**
 * Funções para chamadas de API relacionadas a relatórios
 * Usa o cliente HTTP centralizado
 */

import { api } from './client';

// =============================================================================
// TIPOS
// =============================================================================

export type ReportType = 'market_share' | 'audience_segmentation' | 'event_metrics';
export type ReportFocus = 'general' | 'brands' | 'products' | 'audience';

export interface MarketShareFilters {
  date_from: string;  // YYYY-MM-DD
  date_to: string;    // YYYY-MM-DD
  sport?: string;
  location?: string;
  product_type?: string;
  brands?: string[];
}

export interface AudienceSegmentationFilters {
  date_from: string;
  date_to: string;
  sport?: string;
  location?: string;
  product_type?: string;
}

export interface EventMetricsFilters {
  event_id: string;
  focus?: ReportFocus;
}

export type ReportFilters = MarketShareFilters | AudienceSegmentationFilters | EventMetricsFilters;

export interface GenerateReportRequest {
  type: ReportType;
  filters: ReportFilters;
}

export interface ReportMetadata {
  total_events: number;
  total_athletes: number;
  total_items: number;
  tokens_used: number;
  model: string;
  generation_time_ms: number;
}

export interface GenerateReportResponse {
  type: ReportType;
  generated_at: string;
  filters_applied: Record<string, unknown>;
  title: string;
  content: string;
  metadata: ReportMetadata;
}

export interface ReportStatusResponse {
  available: boolean;
  model?: string;
  message: string;
}

// =============================================================================
// FUNÇÕES DE API
// =============================================================================

/**
 * Gera um relatório usando LLM
 */
export async function generateReport(
  request: GenerateReportRequest
): Promise<GenerateReportResponse> {
  return api.post<GenerateReportResponse>('/api/reports/generate', request);
}

/**
 * Verifica se o serviço de relatórios está disponível
 */
export async function getReportsStatus(): Promise<ReportStatusResponse> {
  return api.get<ReportStatusResponse>('/api/reports/status');
}

// =============================================================================
// HELPERS
// =============================================================================

/**
 * Formata o conteúdo do relatório para exibição
 */
export function formatReportContent(content: string): string {
  // Adiciona quebras de linha para parágrafos
  return content.replace(/\n\n/g, '\n\n').trim();
}

/**
 * Calcula o tempo de geração em formato legível
 */
export function formatGenerationTime(ms: number): string {
  if (ms < 1000) {
    return `${ms}ms`;
  }
  return `${(ms / 1000).toFixed(1)}s`;
}

/**
 * Retorna o nome amigável do tipo de relatório
 */
export function getReportTypeName(type: ReportType): string {
  const names: Record<ReportType, string> = {
    market_share: 'Market Share',
    audience_segmentation: 'Segmentação de Público',
    event_metrics: 'Métricas do Evento',
  };
  return names[type];
}

/**
 * Retorna o nome amigável do foco do relatório
 */
export function getReportFocusName(focus: ReportFocus): string {
  const names: Record<ReportFocus, string> = {
    general: 'Visão Geral',
    brands: 'Marcas',
    products: 'Produtos',
    audience: 'Público',
  };
  return names[focus];
}

File: app/api/reports.py
"""
Rotas de geração de relatórios com LLM
Endpoints para gerar relatórios de Market Share, Segmentação e Métricas de Evento
"""

from fastapi import APIRouter, Depends, HTTPException
from app.api.deps import get_db_service, get_openai_service
from app.services.database import DatabaseService
from app.services.openai_service import OpenAIService
from app.core.reports import ReportsService
from app.schemas.reports import (
    GenerateReportRequest,
    GenerateReportResponse,
    MarketShareFilters,
    AudienceSegmentationFilters,
    EventMetricsFilters,
)

router = APIRouter()


@router.post("/reports/generate", response_model=GenerateReportResponse)
async def generate_report(
    request: GenerateReportRequest,
    db: DatabaseService = Depends(get_db_service),
    openai: OpenAIService = Depends(get_openai_service)
):
    """
    Gera um relatório usando LLM com base nos filtros fornecidos
    
    ## Tipos de Relatório
    
    ### 1. Market Share (`market_share`)
    Análise de participação de mercado das marcas.
    
    **Filtros:**
    - `date_from` (obrigatório): Data inicial (YYYY-MM-DD)
    - `date_to` (obrigatório): Data final (YYYY-MM-DD)
    - `sport`: Filtrar por esporte
    - `location`: Filtrar por localização
    - `product_type`: Filtrar por tipo de produto
    - `brands`: Lista de marcas específicas para analisar
    
    ### 2. Segmentação de Público (`audience_segmentation`)
    Análise do perfil demográfico e preferências.
    
    **Filtros:**
    - `date_from` (obrigatório): Data inicial
    - `date_to` (obrigatório): Data final
    - `sport`: Filtrar por esporte
    - `location`: Filtrar por localização
    - `product_type`: Filtrar por tipo de produto
    
    ### 3. Métricas do Evento (`event_metrics`)
    Resumo executivo de um evento específico.
    
    **Filtros:**
    - `event_id` (obrigatório): UUID do evento
    - `focus`: Foco do relatório (general, brands, products, audience)
    """
    if openai is None:
        raise HTTPException(
            status_code=503,
            detail="Serviço de geração de relatórios não disponível. "
                   "Configure OPENAI_API_KEY no arquivo .env"
        )
    
    try:
        reports_service = ReportsService(db, openai)
        
        if request.type == "market_share":
            filters = MarketShareFilters(**request.filters)
            return reports_service.generate_market_share_report(filters)
        
        elif request.type == "audience_segmentation":
            filters = AudienceSegmentationFilters(**request.filters)
            return reports_service.generate_audience_segmentation_report(filters)
        
        elif request.type == "event_metrics":
            filters = EventMetricsFilters(**request.filters)
            return reports_service.generate_event_metrics_report(filters)
        
        else:
            raise HTTPException(
                status_code=400,
                detail=f"Tipo de relatório inválido: {request.type}. "
                       f"Tipos válidos: market_share, audience_segmentation, event_metrics"
            )
    
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except RuntimeError as e:
        raise HTTPException(status_code=500, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=500, 
            detail=f"Erro ao gerar relatório: {str(e)}"
        )


@router.get("/reports/status")
async def get_reports_status(
    openai: OpenAIService = Depends(get_openai_service)
):
    """
    Verifica se o serviço de relatórios está disponível
    
    Retorna:
    - available: true se OPENAI_API_KEY está configurada
    - model: modelo LLM que será utilizado
    """
    if openai is None:
        return {
            "available": False,
            "message": "OPENAI_API_KEY não configurada"
        }
    
    return {
        "available": True,
        "model": openai.model,
        "message": "Serviço de relatórios disponível"
    }

